File: generated_code/python/solution_838_3.py
from typing import Optional

class ListNode:
    """Standard Node definition for a doubly linked list."""
    def __init__(self, val: int):
        self.val = val
        self.prev: Optional['ListNode'] = None
        self.next: Optional['ListNode'] = None

class MyLinkedList:
    """
    Implementation using a Doubly Linked List with sentinel (dummy) nodes.
    Sentinel nodes at the head and tail simplify edge cases for insertion 
    and deletion, as every 'real' node always has a neighbor.
    
    Complexity:
    - get: O(k) where k is the index.
    - addAtHead/Tail: O(1).
    - addAtIndex/deleteAtIndex: O(k).
    - Space: O(n) to store n elements.
    """

    def __init__(self):
        self.size = 0
        self.head = ListNode(0)  # Sentinel head
        self.tail = ListNode(0)  # Sentinel tail
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, index: int) -> int:
        if index < 0 or index >= self.size:
            return -1
        
        # Optimization: choose the shortest path (from head or tail)
        if index + 1 < self.size - index:
            curr = self.head
            for _ in range(index + 1):
                curr = curr.next
        else:
            curr = self.tail
            for _ in range(self.size - index):
                curr = curr.prev
        return curr.val

    def addAtHead(self, val: int) -> None:
        self.addAtIndex(0, val)

    def addAtTail(self, val: int) -> None:
        self.addAtIndex(self.size, val)

    def addAtIndex(self, index: int, val: int) -> None:
        if index > self.size:
            return
        if index < 0:
            index = 0
        
        # Find predecessor and successor of the new node
        if index < self.size - index:
            pred = self.head
            for _ in range(index):
                pred = pred.next
            succ = pred.next
        else:
            succ = self.tail
            for _ in range(self.size - index):
                succ = succ.prev
            pred = succ.prev
            
        self.size += 1
        to_add = ListNode(val)
        to_add.prev = pred
        to_add.next = succ
        pred.next = to_add
        succ.prev = to_add

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.size:
            return
        
        # Find predecessor and successor of the node to be deleted
        if index < self.size - index:
            pred = self.head
            for _ in range(index):
                pred = pred.next
            succ = pred.next.next
        else:
            succ = self.tail
            for _ in range(self.size - index - 1):
                succ = succ.prev
            pred = succ.prev.prev
            
        self.size -= 1
        pred.next = succ
        succ.prev = pred


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)

File: generated_code/cpp/solution_838_3.cpp
#include <iostream>

// Definition for doubly-linked list node.
struct Node {
    int val;
    Node* prev;
    Node* next;
    Node(int v) : val(v), prev(nullptr), next(nullptr) {}
};

class MyLinkedList {
private:
    int size;
    Node* head;
    Node* tail;

public:
    // Initializes the MyLinkedList object with dummy head and tail nodes to simplify insertions/deletions.
    MyLinkedList() {
        size = 0;
        head = new Node(0);
        tail = new Node(0);
        head->next = tail;
        tail->prev = head;
    }
    
    // Destructor to clean up memory and prevent leaks.
    ~MyLinkedList() {
        Node* curr = head;
        while (curr != nullptr) {
            Node* nextNode = curr->next;
            delete curr;
            curr = nextNode;
        }
    }
    
    // Time Complexity: O(min(index, size - index))
    int get(int index) {
        if (index < 0 || index >= size) {
            return -1;
        }
        
        Node* curr;
        // Optimize search direction based on proximity to head or tail
        if (index < size / 2) {
            curr = head->next;
            for (int i = 0; i < index; ++i) {
                curr = curr->next;
            }
        } else {
            curr = tail->prev;
            for (int i = 0; i < size - 1 - index; ++i) {
                curr = curr->prev;
            }
        }
        return curr->val;
    }
    
    // Time Complexity: O(1)
    void addAtHead(int val) {
        Node* newNode = new Node(val);
        Node* succ = head->next;
        
        newNode->prev = head;
        newNode->next = succ;
        head->next = newNode;
        succ->prev = newNode;
        
        size++;
    }
    
    // Time Complexity: O(1)
    void addAtTail(int val) {
        Node* newNode = new Node(val);
        Node* pred = tail->prev;
        
        newNode->prev = pred;
        newNode->next = tail;
        pred->next = newNode;
        tail->prev = newNode;
        
        size++;
    }
    
    // Time Complexity: O(min(index, size - index))
    void addAtIndex(int index, int val) {
        if (index < 0 || index > size) {
            return;
        }
        
        Node* pred;
        Node* succ;
        
        // Find predecessor and successor for insertion point
        if (index < size / 2) {
            pred = head;
            for (int i = 0; i < index; ++i) {
                pred = pred->next;
            }
            succ = pred->next;
        } else {
            succ = tail;
            for (int i = 0; i < size - index; ++i) {
                succ = succ->prev;
            }
            pred = succ->prev;
        }
        
        Node* newNode = new Node(val);
        newNode->prev = pred;
        newNode->next = succ;
        pred->next = newNode;
        succ->prev = newNode;
        
        size++;
    }
    
    // Time Complexity: O(min(index, size - index))
    void deleteAtIndex(int index) {
        if (index < 0 || index >= size) {
            return;
        }
        
        Node* pred;
        Node* succ;
        
        // Find predecessor and successor of the node to be deleted
        if (index < size / 2) {
            pred = head;
            for (int i = 0; i < index; ++i) {
                pred = pred->next;
            }
            succ = pred->next->next;
        } else {
            succ = tail;
            for (int i = 0; i < size - 1 - index; ++i) {
                succ = succ->prev;
            }
            pred = succ->prev->prev;
        }
        
        Node* toDelete = pred->next;
        pred->next = succ;
        succ->prev = pred;
        
        delete toDelete;
        size--;
    }
};